packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,349 @@
+        if (
+          openRef.current &&
+          dialogRef.current?.returnValue !== 'skipOnClose'
+        ) {
+          onClose?.();
+        }
+        if (dialogRef.current?.returnValue === 'skipOnClose') {
+          dialogRef.current.returnValue = '';
+        }
+      };
+
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore The package is bundled incorrectly
+      dialogPolyfill.registerDialog(dialogElement);
+      dialogElement.style.setProperty(
+        '--dialog-animation-duration',
+        `${animationDuration}ms`,
+      );
+      if (preventClose) {
+        dialogElement.addEventListener('keydown', onPolyfillDialogKeydown);
+      }
+      if (onClose) {
+        dialogElement.addEventListener('close', handleClose);
+      }
+
+      return () => {
+        if (onClose) {
+          dialogElement.removeEventListener('close', handleClose);
+        }
+        if (preventClose) {
+          dialogElement.removeEventListener('keydown', onPolyfillDialogKeydown);
+        }
+      };
+    }, [
+      openRef.current,
+      onClose,
+      preventClose,
+      onPolyfillDialogKeydown,
+      animationDuration,
+    ]);
+
+    const handleDialogClose = useCallback(() => {
+      const dialogElement = dialogRef.current;
+      if (!dialogElement) {
+        return;
+      }
+
+      if (!hasNativeDialog && isModal) {
+        (dialogElement.nextSibling as HTMLDivElement).classList.remove(
+          classes['backdrop-visible'],
+        );
+      }
+      onCloseStart?.();
+      // trigger closing of the dialog after animation
+      setTimeout(() => {
+        if (dialogElement.open) {
+          dialogElement.close();
+        }
+      }, animationDuration);
+    }, [hasNativeDialog, animationDuration, isModal, onCloseStart]);
+
+    useClickOutside([dialogRef], () => handleDialogClose(), open && !isModal);
+    useEscapeKey(() => handleDialogClose(), open && !isModal);
+
+    useEffect(() => {
+      const dialogElement = dialogRef.current;
+
+      if (!dialogElement) {
+        return undefined;
+      }
+
+      if (!open) {
+        dialogElement.returnValue = '';
+      }
+      if (open) {
+        if (!dialogElement.open) {
+          if (isModal) {
+            dialogElement.showModal();
+          } else {
+            dialogElement.show();
+          }
+
+          if (!hasNativeDialog && dialogElement.nextSibling) {
+            // use the polyfill backdrop
+            (dialogElement.nextSibling as HTMLDivElement).classList.add(
+              classes['backdrop-visible'],
+              classes.backdrop,
+            );
+            // intercept and prevent modal closing if preventClose is true
+            (dialogElement.nextSibling as HTMLDivElement).addEventListener(
+              'click',
+              onPolyfillBackdropClick,
+            );
+          }
+        }
+      } else if (dialogElement.open) {
+        handleDialogClose();
+      }
```

Comment:
I find these deeply nested `if`s hard to parse, here's a suggestion to make it a bit more readable: 

```suggestion
      if (open && !dialogElement.open) {
        if (isModal) {
          dialogElement.showModal();
        } else {
          dialogElement.show();
        }

        if (!hasNativeDialog && dialogElement.nextSibling) {
          // use the polyfill backdrop
          (dialogElement.nextSibling as HTMLDivElement).classList.add(
            classes['backdrop-visible'],
            classes.backdrop,
          );
          // intercept and prevent modal closing if preventClose is true
          (dialogElement.nextSibling as HTMLDivElement).addEventListener(
            'click',
            onPolyfillBackdropClick,
          );
        }
      }

      if (!open && dialogElement.open) {
        handleDialogClose();
      }
```